Putting together a draft paper for JOSS (#347)
devosoft/Empirical: paper.md

   index: 6
 - name: Computer Science, Grand Valley State University
   index: 7
date: 30 July 2021
bibliography: paper.bib
breaks: false
---

<!-- [Scraps document here](https://hackmd.io/@DSAbJHksRtqYhjQ4NWzmGA/S1z3oQDWv) -->
<!--  @AML (2021-06-12): Doing a comment clean up, moving important ones (e.g., old meeting notes) into the scraps document. -->

<!-- INSTRUCTIONS

- The paper should be between 250-1000 words.

Paper should include:

- A summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience.
- A clear Statement of Need that illustrates the research purpose of the software.
- A list of key references, including to other software addressing related needs.
- Mention (if applicable) a representative set of past or ongoing research projects using the software and recent scholarly publications enabled by it.
- Acknowledgement of any financial support.

-->

<!-- RESOURCES

- Empirical coding guidelines: https://github.com/devosoft/Empirical/wiki/Coding-Guidelines
- Empirical README: https://github.com/devosoft/Empirical/blob/master/README.md

-->

> # Summary

Reviewer: Summary heading twice?

Reviewer 2: ```suggestion
```

Reviewer 2: ```suggestion
```

2. **Efficiency:** Empirical tools emphasize efficiency to make heavy scientific computing workloads tractable. Where possible, computation is moved to compile-time, heap allocation is avoided, and memory-locality is respected.
3. **Reliability:** Scientific simulation and data analysis is meaningful insofar as it is correct. In conjunction with extensive unit testing, Empirical benefits from --- and provides --- sophisticated debug-mode instrumentation tools. These include audited memory management, drop-in replacements for standard C++ containers with safety checking, and self-documenting assertions.
4. **Distributability:** Scientific software should be easily used by any researcher, student, or citizen scientist who wants to experiment, explore, or replicate studies. As such, Empirical is highly portable, uses common data formats, facilitates flexible runtime configuration, and simplifies compilation to a performant web app.

> 

Reviewer 2: ```suggestion

The core Empirical support library comprises four major components:

 1. debug-instrumented fundamental type and STL container wrappers,
 2. implementations of general-purpose data structures and algorithms,
 3. integrated, end-to-end frameworks for data and configuration management, and
 4. object-oriented bindings for Emscripten/WebAssembly GUI elements.
```

Reviewer 3: if first mention of STL:
"STL" ==> standard template library (STL)

Reviewer 2: per cpp reference, it looks the official name is actually “standard library”! will change and not have acronym

Reviewer 4: Somehow I completely missed that the STL was never official and that the C++ Standard Library was heavily based off of it, but is technically something new with C++ 11.

# Statement of Need
<!-- STATEMENT OF NEED
JOSS: illustrates the research purpose of the software.
JOSS reviewer checklist: "clearly state what problems the software is designed to solve and who the target audience is"
-->

> Modern web-based interfaces give computational research the unique potential to embody open science objectives of making the scientific process more transparent with auditable and exensible code, clear and replicable methodologies, and production of accessible results [CITE osf.io].

Reviewer 3: ```
Modern web-based interfaces give computational research the unique potential to embody open science objectives of making the scientific process more transparent with auditable and exensible code, clear and replicable methodologies, and production of accessible results [CITE osf.io].
```

Don't think there's anything grammatically wrong here, but this would be more readable if broken up. E.g., period after "open science objectives" and make the rest into a new sentence.

Reviewer 2: ```suggestion
Modern web-based interfaces give computational research the unique potential to embody open science objectives.
They can make the scientific process more transparent with auditable and extensible code, clear and replicable methodologies, and production of accessible results [CITE osf.io].
```

<!-- [@world_academy_of_science_engineering_and_technology_2019] -->

> In practice, however, many scientific software applications are difficult to obtain, install, or use, and produce data in propriatary formats.

Reviewer 2: ```suggestion
In practice, however, many scientific software applications are difficult to obtain, install, or use, and produce data in proprietary formats.
```

> High quality open-science tools encourage researchers to follow effective software development practices by simplifying development and helping them improve code quality, scientific rigor, and ease of replication or extension, resulting in broader confidence in and adoption of their software.

Reviewer 3: ```
High quality open-science tools encourage researchers to follow effective software development practices by simplifying development and helping them improve code quality, scientific rigor, and ease of replication or extension, resulting in broader confidence in and adoption of their software.
```

Breaking this sentence up would improve readability. E.g., period after "extension", and put the "resulting" clause in a new sentence.

Reviewer 2: ```suggestion
High quality open-science tools encourage researchers to follow effective software development practices by simplifying development and helping them improve code quality, scientific rigor, and ease of replication or extension.
```

In the process, researchers get more out of their own code.

> For example, an effective toolkit would allow a developer to easily add a GUI that would not only help other users of the software, but could also help the researcher themsleves gain "soft knowledge" of their system [@10.1145/3185517].

Reviewer: I would cut "an effective toolkit would allow a developer to easily add a" and rewrite to just "For example, adding a GUI not only helps other uses of the software, but could also help..."

* building the GUI off of highly-efficient Web Assembly, and
* computing everything possible at compile time to increase execution efficiency.
 -->
<!-- Even with an easily available web-based interface, scientific software cannot fulfill open science objectives if its credbility is compromised by undiscovered errors.
Determining the presence of subtle runtime errors in native executables generated using C++ can be difficult, and debugging compiled WASM output from Emscripten can be nearly impossible.
Tools to support researchers in detecting and correcting such errors are critical to open science objectives. -->
<!-- To improve researchers' ability to debug their software, Empirical provides a number of improved debugging functionalities, including wrappers around standard library containers that include extra safety checks, smart pointers in debug mode that convert to raw pointers in release mode, and improved assertions.
These safeguards can be fully disabled by a compiler flag in order to ensure maximum efficiency in performance-critical contexts.
 -->

# Empirical Features

## Facilitating Better Code for Scientific Software

> Software produced by academics, especially for one-off use often foregoes rigorous programming practices for the sake of expediency.

Reviewer: Insert comma between "use" and "often"

Reviewer 2: ```suggestion
Software produced by academics, especially for one-off use, often foregoes rigorous programming practices for the sake of expediency.
```

## Projects Using the Software

* [Aagos](https://lalejini.com/Aagos/web/Aagos.html) [@gillespie2018changing]
    * An interactive model for empirically studying how evinronmental change affects the evolution of genetic architectures.
* [conduit](https://conduit.fyi) [@moreno2020conduit]
  * A C++ library that wraps intra-thread, inter-thread, and inter-process communication in a uniform, modular, object-oriented interface, with a focus on asynchronous high-performance computing applications.
* [Dishtiny](https://mmore500.com/dishtiny) [@moreno2019toward]
  * An interactive web demo of an artificial life platform built to study major transitions in evolution.
* [ecology in evolutionary computation explorer](https://emilydolson.github.io/ecology_of_evolutionary_computation/web/interaction_networks.html) [@dolson2018evological]
  * Interactive exploration of interaction networks in evolutionary computation under different selection schemes.
* [Symbulation](https://anyaevostinar.github.io/SymbulationEmp/web/symbulation.html) [@vostinar2017suicide]
  * An interactive artificial life model focused on the evolution of symbiosis between parasitism to mutualism.

>     Also features a game to recreate major results from published work.

Reviewer: Game no longer in place, can remove this sentence :)

Reviewer 2: ```suggestion
```
